### PR TITLE
Support nested types introspection in C/C++

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says. While working on typesupport introspection tests, I discovered that (a) the API for basic type arrays' introspection in C was completely missing (#598), and (b) the API for bool arrays' introspection in C++ was missing (and arguably broken due to `std::vector<bool>` quirks). 

This patch amends this by, among other things, introducing an `std::vector` subclass with its own specialization for `bool`. Note that since `std::vector<bool>` does not necessarily use contiguous storage, it makes type erased access to its elements via `void *` outright impossible (and that's exactly what introspection APIs do).

### CODE REVIEW

**Comment by j-rivero** [comment]
Could the `#include <vector>` be removed from here https://github.com/ros2/rosidl/pull/647/files#diff-a05f4acc14bbcd36da0dfc3e0b606020ec48d5294b58f06b3cf9f9e593aa3cf1R21? 

A totally undocumented questions: the same that happen for `std::vector` could happen for `std::array`? There is one https://github.com/ros2/rosidl/blob/master/rosidl_generator_cpp/rosidl_generator_cpp/__init__.py#L130.

File: rosidl_generator_cpp/resource/idl__struct.hpp.em [hunk 1]
```diff
@@ -29,6 +29,7 @@ include_directives = set()
 #include <string>
 #include <vector>
```
Comment (j-rivero) [inline rosidl_generator_cpp/resource/idl__struct.hpp.em:30]: Would the include of vector now be superfluous?

**Review by j-rivero — APPROVED** [review]
I can not detect any error, although my knowledge of the package is very limited. A couple of minor questions before merging thought.

File: rosidl_generator_cpp/resource/idl__struct.hpp.em (hunk 1 above)
Comment (hidmic) [reply rosidl_generator_cpp/resource/idl__struct.hpp.em:30]: Good point.

**Comment by hidmic** [comment]
> A totally undocumented questions: the same that happen for std::vector could happen for std::array? There is one https://github.com/ros2/rosidl/blob/master/rosidl_generator_cpp/rosidl_generator_cpp/__init__.py#L130.

@j-rivero fortunately, no. Only `std::vector<bool>` is special in an awkward sense.

**Review by gbiggs — APPROVED** [review]

**Comment by hidmic** [comment]
Full CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15987/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10657/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16309/)

File: rosidl_runtime_cpp/include/rosidl_runtime_cpp/vector.hpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+  using Base::vector;
+
+  // Implicit "copy" and "move" constructors from std::vector<bool> instances
+  Vector(const std::vector<bool, AllocatorT> & other)  // NOLINT(runtime/explicit)
+  : Base(other.begin(), other.end())
+  {
+  }
+
+  Vector(std::vector<bool, AllocatorT> && other)  // NOLINT(runtime/explicit)
```
Comment (ivanpauno) [inline rosidl_runtime_cpp/include/rosidl_runtime_cpp/vector.hpp:149]: I think we should keep this implicit conversions, to make the transition smooth, but they are really bad performance wise.
IMHO, things that are costly should only be possible in an explicit way.

So I think we should add the implicit conversions with a deprecatd message, and provide explicit alternatives (e.g. a static method `from_std_vector_bool()`).
I don't mind much though.

Comment (hidmic) [reply rosidl_runtime_cpp/include/rosidl_runtime_cpp/vector.hpp:149]: > I think we should keep this implicit conversions, to make the transition smooth, but they are really bad performance wise.

Yeah, I didn't want to cause breakage downstream but you're right.

> So I think we should add the implicit conversions with a deprecatd message, and provide explicit alternatives

That sounds good to me. Some `Vector::assign` overload and/or making constructors explicit would do too.

**Comment by hidmic** [comment]
@ivanpauno @j-rivero I'm rolling back with this. Too many C++ template partial specializations that work with `std::vector` are breaking downstream (and it doesn't seem like I can seamlessly bend template resolution rules to stop that from happening). 

I'll re-open this PR with just the C stuff. We'll have to live with `std::vector<bool>`.

**Comment by ivanpauno** [comment]
> Too many C++ template partial specializations that work with std::vector are breaking downstream

I'm curious what those look like.
Maybe introducing a `VectorBoolContinuous` and continue using `std::vector<T>` for everything else reduces the ammount of code being break in downstream packages.

Though that's maybe out of scope of what you're trying to do.

**Pull request closed without merging**